Auto-Finding Servers on LAN

LMiljak opened the pull request:
Added two complicated classes:
  -ServerFinder class
  -ClientFinder class

They are both used for a client to easily find a server on the local network. These two classes are heavily inspired by the following article:
http://michieldemey.be/blog/network-discovery-using-udp-broadcast/
, so I highly recommend reading that first before this code. Don't worry it's a short article.

Take your time reviewing the code and asking questions as the code uses both MultiThreading as well as Sockets, which  some of us aren't fully familiar with.

I decided to not add test cases, because when I tried, it was chaotic. Testing multiplayer and sockets is next to impossible.

So because there aren't any unit test, you should test this manually. I've added two main methods in both the ClientFinder and the ServerFinder classes. The ClientFinder is meant to be run by the server and the ServerFinder is meant to be run by the client (don't confuse them). You can be both client and server on the same computer. To do this simply run the main in ClientFinder on one window and the main in ServerFinder in another window. If both main methods are running, you should see in the console of the ServerFinder that it has found a server (every 5 seconds) together with the address of the server.
If you have another computer in your LAN, you can try and run one main method on that device, and the other on your current one.
Pressing the enter key in one of the consoles terminates the Client/ServerFinder, just make sure to terminate the other one as well, or you might have it running forever without you noticing.

Also, if you want to test this with another computer on your LAN, but don't have an IDE/Github/etc on it, but you do have Java installed. Then you can (from your computer that does have it) right-click on the project (from Eclipse) -> export -> Java -> Runnable JAR file. Select the correct main method in the Launch configurations (either from the ClientFinder or ServerFinder), and export it (ignore any warnings). Send the jar file to the other computer, then open it from the command line using "java -jar filename.jar". You can then launch the other main method from your other computer from your IDE, and hope it works.

Commit:
An example of a simple server-client connection.

This example is pretty much copy & pasted from
https://wiki.jmonkeyengine.org/doku.php/jme3:advanced:networking
I didn't add tests for this is just a small example, it will be removed
soon enough.

Commit:
Added a sending Messages to the example networking.

Created a "PrintMessage" class, which when received by the client or
server, prints the message to the console. The ClientListener and
ServerListener classes handle these messages.

Commit:
Added basic implementation for a the host.

The host acts as a kind of wrapper class for the actual server.

Commit:
A Discoverer class that helps client find a server.

The responsibility of this class is heavily inspired by the following
article:
http://michieldemey.be/blog/network-discovery-using-udp-broadcast/
I certainly recommend quickly reading through it before code reviewing.

Commit:
Renamed some packages.

So it fits all naming conventions.

Commit:
Removed the example networking package.

Commit:
Removed the example test generated by Maven.

Commit:
Added some method to the Host class

Added a startServer and a closeServer method. The server no longer
starts when the private constructor is called.

Commit:
Added some tests and fixed bugs in host class

Commit:
Removed the HostTest completely.

I've concluded that testing the networking package is next to
impossible.

Commit:
Renamed Discoverer to ClientFinder

Makes a little bit more sense because the Discoverer only waits for
packets to be send to him instead of trying to discover things itself.

Commit:
Some extra javadoc and changed a method name

handleMessages has now been rename to findClients.
Added some extra Javadoc to explain the purpose of the method.

Commit:
ClientFinder now runs on a new Thread

- The public findClients method now requires an executorService as
parameter.
- Made the socket run on a different port than the Host class.
- There is now a stop method that stops running the findClients task and
closes the socket it was running on.

Commit:
Moved class fields above a method.

Commit:
Fixed some Javadoc errors in ClientFinder.

Commit:
Added the ServerFinder class.

It is the lost brother of the ClientFinder class.
See http://michieldemey.be/blog/network-discovery-using-udp-broadcast/
for explanations. This code is highly based on the article.

Commit:
Added main method in the Client&ServerFinder

These main methods should be used for testing purposes only.

Commit:
Temporary removed the Host class

A pull request is coming up soon, and the Host class would serve no
purpose as the main new features are the Client&ServerFinders. The Host
class will come back up after it.

Commit:
Maven now compiles using Java 1.8

Commit:
Fixed bug where LAN-devices won't receive a packet.

MrCrime commented:
Also, if you want to test this with another computer on your LAN, but don't have an IDE/Github/etc on it, but you do have Java installed. Then you can (from your computer that does have it) right-click on the project (from Eclipse) -> export -> Java -> Runnable JAR file. Select the main method you want to export (either from the ClientFinder or ServerFinder), and export it (ignore any warnings). Send the jar file to the other computer, then open it from the command line using "java -jar <jarname>.jar". You can then launch the other main method from your other computer from your IDE, and hope it works.

NilsHullegien commented:
Looks good and tested!

Commit:
Merge remote-tracking branch 'refs/remotes/origin/master' into networking

Commit:
Merge remote-tracking branch 'refs/remotes/origin/master' into networking

Commit:
Changed location of the classes

Moved the classes to new packages so it fits with the adjusted
achitecture design
